File: notes-new-drawing/app/src/main/java/com/wyldsoft/notes/shapemanagement/shapes/TextShape.java
package com.wyldsoft.notes.shapemanagement.shapes;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;

import com.onyx.android.sdk.data.note.TouchPoint;
import com.onyx.android.sdk.pen.data.TouchPointList;
import com.wyldsoft.notes.rendering.RendererHelper;

import java.util.List;

public class TextShape extends BaseShape {

    private String text = "";
    private float fontSize = 32f;
    private String fontFamily = "sans-serif";

    public void setText(String text) {
        this.text = text != null ? text : "";
    }

    public String getText() {
        return text;
    }

    public void setFontSize(float fontSize) {
        this.fontSize = fontSize;
    }

    public float getFontSize() {
        return fontSize;
    }

    public void setFontFamily(String fontFamily) {
        this.fontFamily = fontFamily != null ? fontFamily : "sans-serif";
    }

    public String getFontFamily() {
        return fontFamily;
    }

    private Typeface resolveTypeface() {
        switch (fontFamily) {
            case "serif":
                return Typeface.SERIF;
            case "monospace":
                return Typeface.MONOSPACE;
            case "sans-serif":
            default:
                return Typeface.SANS_SERIF;
        }
    }

    @Override
    public void render(RendererHelper.RenderContext renderContext) {
        if (text.isEmpty() || touchPointList == null) return;

        List<TouchPoint> points = touchPointList.getPoints();
        if (points == null || points.isEmpty()) return;

        Canvas canvas = renderContext.canvas;
        Paint paint = renderContext.paint;

        paint.reset();
        paint.setColor(strokeColor);
        paint.setTextSize(fontSize * renderContext.viewportScale);
        paint.setTypeface(resolveTypeface());
        paint.setAntiAlias(true);
        paint.setStyle(Paint.Style.FILL);

        float x = points.get(0).x;
        float y = points.get(0).y;

        canvas.drawText(text, x, y, paint);
    }

    @Override
    public void updateShapeRect() {
        if (text.isEmpty() || touchPointList == null) {
            originRect = new RectF(0, 0, 0, 0);
            boundingRect = new RectF(originRect);
            return;
        }

        List<TouchPoint> points = touchPointList.getPoints();
        if (points == null || points.isEmpty()) {
            originRect = new RectF(0, 0, 0, 0);
            boundingRect = new RectF(originRect);
            return;
        }

        float x = points.get(0).x;
        float y = points.get(0).y;

        Paint measurePaint = new Paint();
        measurePaint.setTextSize(fontSize);
        measurePaint.setTypeface(resolveTypeface());
        Rect textBounds = new Rect();
        measurePaint.getTextBounds(text, 0, text.length(), textBounds);

        originRect = new RectF(
                x + textBounds.left,
                y + textBounds.top,
                x + textBounds.right,
                y + textBounds.bottom
        );
        boundingRect = new RectF(originRect);
    }

    @Override
    public boolean hitTestPoints(TouchPointList pointList, float radius) {
        updateShapeRect();
        if (boundingRect == null) return false;
        RectF expanded = new RectF(
                boundingRect.left - radius,
                boundingRect.top - radius,
                boundingRect.right + radius,
                boundingRect.bottom + radius
        );
        for (TouchPoint tp : pointList.getPoints()) {
            if (expanded.contains(tp.x, tp.y)) return true;
        }
        return false;
    }
}

File: notes-new-drawing/app/src/main/java/com/wyldsoft/notes/shapemanagement/shapes/TextShape.kt
package com.wyldsoft.notes.shapemanagement.shapes

import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.util.Log
import com.wyldsoft.notes.rendering.RenderContext

/**
 * A shape that renders a text string at a fixed position on the canvas.
 * Position is stored as a single-point TouchPointList (x, y in note coordinates).
 * strokeWidth holds the font size in pixels; strokeColor holds the text color.
 */
class TextShape : Shape() {
    companion object {
        private const val TAG = "TextShape"
    }

    var text: String = ""
    var fontFamily: String = "sans-serif"

    override fun render(renderContext: RenderContext) {
        Log.d(TAG, "render text='$text' font=$fontFamily")
        val canvas = renderContext.canvas ?: return
        val pt = touchPointList?.points?.firstOrNull() ?: return
        val typeface = when (fontFamily) {
            "serif" -> Typeface.SERIF
            "monospace" -> Typeface.MONOSPACE
            else -> Typeface.SANS_SERIF
        }
        val paint = Paint().apply {
            isAntiAlias = true
            color = if (strokeColor != 0) strokeColor else Color.BLACK
            textSize = if (strokeWidth > 0f) strokeWidth else 32f
            this.typeface = typeface
            style = Paint.Style.FILL
        }
        canvas.drawText(text, pt.x, pt.y, paint)
    }
}
